File: #120_Triangle/Solution.java
import java.util.*;

public class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        if (triangle == null || triangle.size() == 0) return 0;
        int n = triangle.size();
        int[] dp = new int[n + 1];
        for (int i = 0; i < n + 1; i++) dp[i] = Integer.MAX_VALUE;
        for (List<Integer> row : triangle) {
        	if (row.size() == 1) {
        		dp[1] = row.get(0);
        	}
        	else {
        		int index = 1;
        		int prev = dp[index - 1];
        		for (Integer it : row) {
        			int tmp = dp[index];
        			dp[index] = Math.min(dp[index], prev) + it;
        			prev = tmp;
        			index++;
        		}
        	}
        }
        int min = Integer.MAX_VALUE;
        for (Integer it : dp) {
        	min = Math.min(it, min);
        }
        return min;
    }
}

File: #120_Triangle/Solution.py
import sys

class Solution:
    # @param triangle, a list of lists of integers
    # @return an integer
    def minimumTotal(self, triangle):
        if not triangle:
        	return 0
        dp = [0 for j in range(len(triangle))]
        for x in range(len(triangle)):
        	for y in range(x + 1):
        		if x == 0:
        			ul, ur = 0, 0
        		elif y == 0:
        			ul = sys.maxint
        			ur = dp[y]
        		elif y == x:
        			ul = tmp
        			ur = sys.maxint
        		else:
        			ul = tmp
        			ur = dp[y]
        		tmp2 = dp[y]
        		dp[y] = triangle[x][y] + min(ul, ur)
        		tmp = tmp2
        return min(dp)

sol = Solution()
tri = [[2], [3, 4], [6, 5, 7], [4, 1, 8, 3]]
print sol.minimumTotal(tri)